app/logic/Mail/ActiveSync/ActiveSyncAccount.ts

```diff
@@ -452,41 +442,53 @@ export class ActiveSyncAccount extends MailAccount {
             await folder.save();
             missingFolders.remove(folder);
             break;
-          case FolderType.Contacts:
-          case FolderType.UserContacts:
-            let haveAddressbook = appGlobal.addressbooks.find(addressbook => addressbook.mainAccount == this);
-            let isMainAddressbook = !haveAddressbook; // TODO Determine default addressbook
-            if (!haveAddressbook) {
-              let addressbook = newAddressbookForProtocol("addressbook-activesync") as ActiveSyncAddressbook;
-              addressbook.initFromMainAccount(this);
-              if (!isMainAddressbook) {
-                addressbook.name = sanitize.nonemptylabel(change.DisplayName, addressbook.name);
-              }
-              addressbook.url = sanitize.url(folderURL);
-              addressbook.serverID = sanitize.nonemptystring(change.ServerId);
-              await addressbook.save();
-              appGlobal.addressbooks.add(addressbook);
-            }
+          case FolderType.Tasks:
+          case FolderType.UserTasks:
+            // Mustang doesn't support tasks yet, fortunately.
             break;
           case FolderType.Calendar:
           case FolderType.UserCalendar:
-            let haveCalendar = appGlobal.calendars.find(calendar => calendar.mainAccount == this);
-            let isMainCalendar = !haveCalendar; // TODO Determine default calendar
-            if (!haveCalendar) {
-              let calendar = newCalendarForProtocol("calendar-activesync") as ActiveSyncCalendar;
-              calendar.initFromMainAccount(this);
-              if (!isMainCalendar) {
-                calendar.name = sanitize.nonemptylabel(change.DisplayName, calendar.name);
-              }
-              calendar.url = sanitize.url(folderURL);
-              calendar.serverID = sanitize.nonemptystring(change.ServerId);
-              await calendar.save();
+            let calendar = appGlobal.calendars.find((calendar): calendar is ActiveSyncCalendar => calendar.mainAccount == this && (calendar.url == url.toString() || (calendar as ActiveSyncCalendar).serverID == change.ServerId));
```

Comment:
```suggestion
            let calendar = appGlobal.calendars.find((calendar: ActiveSyncCalendar): calendar is ActiveSyncCalendar => calendar.mainAccount == this && (calendar.url == url.toString() || calendar.serverID == change.ServerId));
```

Does this also work?
And: Is there not a shorter (less redundant) way to express the same?